test/extended/machines/machines.go

```diff
@@ -100,6 +105,37 @@ var _ = g.Describe("[sig-cluster-lifecycle][Feature:Machines] Managed cluster sh
 			e2e.Failf("Machine resources missing for nodes: %s", strings.Join(nodeNames.List(), ", "))
 		}
 	})
+
+	g.It("[sig-scheduling][Early] control plane machine set operator should not cause an early rollout", func() {
+		cfg, err := e2e.LoadConfig()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		c, err := e2e.LoadClientset()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		dc, err := dynamic.NewForConfig(cfg)
+		o.Expect(err).NotTo(o.HaveOccurred())
+		machineClientSet, err := machineclient.NewForConfig(oc.KubeFramework().ClientConfig())
+		o.Expect(err).ToNot(o.HaveOccurred())
+
+		pattern := `^([a-zA-Z0-9]+-)+master-\d+$`
+
+		g.By("checking for the openshift machine api operator")
+		// TODO: skip if platform != aws
+		skipUnlessMachineAPIOperator(dc, c.CoreV1().Namespaces())
+
+		g.By("ensuring every node is linked to a machine api resource")
+		allControlPlaneMachines, err := machineClientSet.MachineV1beta1().Machines("openshift-machine-api").List(context.Background(), metav1.ListOptions{
+			LabelSelector: masterMachineLabelSelector,
+		})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		regex, err := regexp.Compile(pattern)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		for _, m := range allControlPlaneMachines.Items {
```

Comment:
I'm also concerned about a machine getting deleted and then recreated before this test starts.  How could I detect that?  And a machine potentially getting created and then deleted before the test starts. 

Is there an event produced when a new master is observed and when an existing master is deleted.  If not, can we add one so that we can easily detect that situation?

Reply:
If a machine gets deleted and then recreated by the CPMS before this test starts, I think, that this case shouldn't be an issue as we get the list of all control plane machines and the one that is recreated will not match the regex used.
The original machine have name in format `<cluster-name>-master-<index>`, but the recreated one will have `<cluster-name>-master-<random>-<index>` and the test will detect that.

An event is not produced by the CPMS when such rollout happens as far as I know. cc/ @JoelSpeed

Reply:
As Daniel says, when the installer creates the machines, it names the in a predictable format, such as `jspeed-abcde-master-0`, and then the CPMS also uses a predictable format, but it's slightly different, `jspeed-abcde-master-fghij-0` would be a replacement for example. So we can easily detect based on the naming that a CPMS has taken action.

CPMS won't delete a control plane machine until the new machine is ready, so, if it is attempting a replacement, we may see a fourth Machine, but that will have the new name format and should fail this test.

Adding events is definitely possible, we don't have that currently but, it would be easy to add and, would allow us to see rapid delete and create through events which would also give a good signal that something is broken.

@odvarkadaniel do you think you could raise a PR to add the create and delete events? They'll basically contain information similar to the logging for [successful create](https://github.com/openshift/cluster-control-plane-machine-set-operator/blob/main/pkg/machineproviders/providers/openshift/machine/v1beta1/provider.go#L540) and [successful delete](https://github.com/openshift/cluster-control-plane-machine-set-operator/blob/edb6246a9b0e9a250d9a66475f21b4830e60ba5a/pkg/machineproviders/providers/openshift/machine/v1beta1/provider.go#L623). I think generating the events against the CPMS would be suitable, since if we put them against the machines, then it will be harder to collate the events later

Reply:
Can a second `[Early]` test be written to ensure there were no deletes and exactly the right number of creates?  I like having this current test, but I think it's also important to check to be sure the controller's intermediate states were accurate.

Reply:
For CPMS, there should be no Creates or Deletes at present, so zero events for either is the expectation in a normal cluster run where we are not deliberately exercising the CPMS

Reply:
Is it even necessary to create a new test that essentially tests the same thing as the one I have introduced? We know that the name of the rolled out machine will not correspond to the format the regex enforces. Also, as Joel is saying, the CPMS won't have any events during the installation (unless the rollout happens even though it should not) - but we would already know about the failure, as this test would have probably failed as well.